os/drivers/pm/pm.c

```diff
@@ -0,0 +1,204 @@
+     	* Description: Lock PM in PM_NORMAL state by invoking pm_stay().
+     	****************************************************************************/
+    	case PMIOC_LOCK:
+        	/* Always lock PM_NORMAL state, check again whether this implementation
+        	is enough to handle, if not, consider below methods...
+        	1a. PM_ACTIVITY is invoked in some of the driver, if application utlized it, it should draw state back to PM_NORMAL
+        	1b. If 1a cannot cover all condition, add one more syscall api, as a bridge for application to call pm_activity
+        	2a. Else, we can consider to lock/unlock for all state other than PM_SLEEP, ie. fetch state: pm_querystate(PM_IDLE_DOMAIN)
+        	2b. Need to verify if the transition is going to happen, but lock state appears suddenly, what will happen?
+        	*/
+        	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+        	fvdbg("State locked!\n");
+        	if (arg > 0) {
+            		pm_set_timer(PM_LOCK_TIMER, arg);
+        	} else {
+            		g_pm_timer.timer_type = PM_NO_TIMER;
+        	}
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Unlock PM from PM_NORMAL state by invoking pm_relax(). 
+         ****************************************************************************/
+        case PMIOC_UNLOCK:
+            	pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+            	fvdbg("State unlocked!\n");
+            	g_pm_timer.timer_type = PM_NO_TIMER;
+            	ret = OK;
+            	break;
+        /****************************************************************************
+         * Description:
+         *  Make the board wake up after given time (arg) when board sleeps.
+         ****************************************************************************/
+        case PMIOC_WAKEUP_TIMER:
+        	if (g_pm_timer.timer_type == PM_NO_TIMER && arg > 0) {
+        		pm_set_timer(PM_WAKEUP_TIMER, arg);
+        	}	 
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Sets timer interval for next sleep cycle after sending of wifi keep 
+         *  alive signal from UART. It also stores the current time tick 
+         *  in g_pm_timer.last_wifi_alive_send_time that will be used to calculate the 
+         *  duration of required sleep later just before sleep. If another
+         *  interrupt occurs during this interval, the system will still
+         *  wake up.
+         *  Also we dont want to do pm_set_timer to PM_WAKEUP_TIMER if the
+         *  system is already in PM_LOCK_TIMER state. If system is in 
+         *  PM_LOCK_TIMER state, then power_unlock_write can only change that state.
+         *
+         ****************************************************************************/
+        case PMIOC_SET_WIFI_WAKEUP:
```

Comment:
I don't agree the name. PM should not have some module specific name. If you add other periodic timer, will you add that with different ioctl command?